### PR TITLE
[WIP] Port trace-redisplay

### LINKED ISSUE #1014: Port trace-redisplay
```c
DEFUN ("trace-redisplay", Ftrace_redisplay, Strace_redisplay, 0, 1, "P",
       doc: /* Toggle tracing of redisplay.
With ARG, turn tracing on if and only if ARG is positive.  */)
  (Lisp_Object arg)
{
  if (NILP (arg))
    trace_redisplay_p = !trace_redisplay_p;
  else
    {
      arg = Fprefix_numeric_value (arg);
      trace_redisplay_p = XINT (arg) > 0;
    }

  return Qnil;
}
```

### PR DESCRIPTION
**Pull request by liff**
Fixes #1014.

Added Rust features for all of the Emacs checking flags while at it. Let me know if they are not (going to be) needed and I’ll just drop them.

### CODE REVIEW

File: rust_src/src/xdisp.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+/// Toggle tracing of redisplay.
+/// With ARG, turn tracing on if and only if ARG is positive.
+#[cfg(feature = "glyph-debug")]
+#[lisp_fn(min = "0", intspec = "P")]
+pub fn trace_redisplay(arg: Option<EmacsInt>) {
+    unsafe {
+        match arg {
+            None => trace_redisplay_p = !trace_redisplay_p,
+            Some(n) if n > 0 => trace_redisplay_p = true,
```
Comment (shaleh) [inline rust_src/src/xdisp.rs]: While the guard is interesting, could this not be written more simply as `Some(n) => trace_redisplay_p = n > 0,`?

File: rust_src/src/xdisp.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+use remacs_macros::lisp_fn;
+
+use crate::remacs_sys::{trace_redisplay_p, EmacsInt};
+
+/// Toggle tracing of redisplay.
+/// With ARG, turn tracing on if and only if ARG is positive.
+#[cfg(feature = "glyph-debug")]
+#[lisp_fn(min = "0", intspec = "P")]
+pub fn trace_redisplay(arg: Option<EmacsInt>) {
```
Comment (shaleh) [inline rust_src/src/xdisp.rs]:
```suggestion
pub fn trace_redisplay(arg: InteractiveNumericPrefix) {
```

We have a type called `InteractiveNumericPrefix` which replaces what used to be `Fprefix_numeric_value`.

File: rust_src/src/xdisp.rs (hunk 1 above)
Comment (liff) [reply rust_src/src/xdisp.rs]: Indeed, fixed, thanks.

File: rust_src/src/xdisp.rs (hunk 2 above)
Comment (liff) [reply rust_src/src/xdisp.rs]: Changed. Should now match original function more closely (adjusted tests).

**Comment by liff** [comment]
I added some ad-hoc logic to deal with exporting conditionally compiled lisp functions. It does very simple parsing of `#[cfg` lines and so may lead to surprises if one tries to use anything more complicated than `#[cfg(feature = "abc")]`.

Any suggestions for better ways of handling that?

File: rust_src/build.rs [hunk 3]
```diff
@@ -127,9 +127,15 @@ impl<'a> ModuleParser<'a> {
     }
 
     pub fn run(&mut self, in_file: impl BufRead) -> Result<ModuleData, BuildError> {
```
Comment (shaleh) [inline rust_src/build.rs]: Can this be extended to include clippy warnings as well? Because I do not think they work currently.

**Comment by shaleh** [comment]
We should update one of the Travis builds to enable the compiler option. Edit the .travis.yml as part of this commit and it will immediately apply.

**Comment by shaleh** [comment]
This all looks really good @liff

File: rust_src/build.rs (hunk 3 above)
Comment (liff) [reply rust_src/build.rs]: I don’t quite understand what you mean. Can you elaborate a bit?

**Comment by liff** [comment]
> We should update one of the Travis builds to enable the compiler option. Edit the .travis.yml as part of this commit and it will immediately apply.

You mean add `--enable-checking=glyps` for all current jobs or add new jobs with that flag set?

**Comment by liff** [comment]
Came up with slightly different (better?) way of handling the `cfg` attributes. Now the build script just captures them as-is and passes them through to the exported entries.

File: test/rust_src/src/xdisp-tests.el [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+(require 'ert)
+
+(ert-deftest xdisp-tests--trace-redisplay-base ()
+  "Check (trace-redisplay) base cases.
+   Only run when configured with --enable-checks=glyphs or equivalent."
+  (when (fboundp 'trace-redisplay)
+    (should-error (trace-redisplay 1 2))
+    (should-error (trace-redisplay nil 2))
+
+    (should (eq nil (trace-redisplay)))
```
Comment (nickdrozd) [inline test/rust_src/src/xdisp-tests.el]: You can use `(should-not ...)` instead of `(should (eq nil ...))`.

**Comment by shaleh** [comment]
> > We should update one of the Travis builds to enable the compiler option. Edit the .travis.yml as part of this commit and it will immediately apply.
> 
> You mean add `--enable-checking=glyps` for all current jobs or add new jobs with that flag set?

Maybe pick one of the current tests and include it there?

**Comment by nickdrozd** [comment]
What's left on this one? It looks like the `glyph-debug` stuff is what https://github.com/remacs/remacs/pull/1366 needs.

**Comment by brotzeit** [comment]
This looks really good...

**Comment by brotzeit** [comment]
@liff Thanks!

**Pull request merged**